plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+        GetBucketLifecycleConfigurationResponse response = s3Client.getBucketLifecycleConfiguration(request);
+
+        verify(response.rules().stream().anyMatch(
+                rule -> rule.expiration() != null &&
+                        rule.abortIncompleteMultipartUpload() != null &&
+                        rule.status().equals(ExpirationStatus.ENABLED) &&
+                        rule.filter().equals(LifecycleRuleFilter.builder().build())
+                ), "Expected file expiration and abortIncompleteMultipartUpload lifecycle rule for exchange bucket %s", baseDirectory.toString());
+    }
+
+    @Override
+    public void createDirectories(URI dir)
+            throws IOException
+    {
+        // no need to do anything for S3
+    }
+
+    @Override
+    public SliceInput getSliceInput(URI file, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        GetObjectRequest.Builder getObjectRequestBuilder = GetObjectRequest.builder()
+                .bucket(getBucketName(file))
+                .key(keyFromUri(file));
+        S3RequestUtil.configureEncryption(secretKey, getObjectRequestBuilder);
+
+        try {
+            return new InputStreamSliceInput(s3Client.getObject(getObjectRequestBuilder.build(), ResponseTransformer.toInputStream()));
+        }
+        catch (AwsServiceException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public ExchangeStorageWriter createExchangeStorageWriter(URI file, Optional<SecretKey> secretKey)
+    {
+        String bucketName = getBucketName(file);
+        String key = keyFromUri(file);
+
+        return new S3ExchangeStorageWriter(s3Client, s3AsyncClient, bucketName, key, multiUploadPartSize, secretKey);
+    }
+
+    @Override
+    public boolean exists(URI file)
+            throws IOException
+    {
+        // Only used for commit marker files and doesn't need secretKey
+        return getHeadObjectResponse(file, Optional.empty()) != null;
+    }
+
+    @Override
+    public void createEmptyFile(URI file)
+            throws IOException
+    {
+        PutObjectRequest request = PutObjectRequest.builder()
+                .bucket(getBucketName(file))
+                .key(keyFromUri(file))
+                .build();
+
+        try {
+            s3Client.putObject(request, RequestBody.empty());
+        }
+        catch (AwsServiceException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public void deleteRecursively(URI uri)
+    {
+        deleteExecutor.submit(() -> deleteRecursivelySync(uri));
+    }
+
+    @Override
+    public Stream<URI> listFiles(URI dir)
+    {
+        List<S3Object> objects = listS3Directory(dir).contents();
+        return objects.stream().filter(object -> !object.key().endsWith(PATH_SEPARATOR)).map(object -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + object.key(), dir.getFragment());
```

Comment:
can you use `dir.resolve()` here?

Reply:
`dir` is like `s3n://{bucketName}/20211224_001743_00000_93qp9.1/0/0/`, `key` is like `20211224_001743_00000_93qp9.1/0/0/0.data`, we can't directly `resolve` here.